File: python/veval/_sdk.py
from __future__ import annotations
import uuid
from datetime import datetime, timezone
from typing import Any, Callable, Awaitable, Optional, TypeVar

from ._options import VevalOptions
from ._context import VevalExecutionContext
from ._step import Step
from ._tracing import (
    TraceData, SnapshotData, SnapshotDiff, compare_snapshots,
    ReplayOptions, ReplayResult,
)
from ._assertions import ITraceAssertion
from ._scenarios import ScenarioItem, ScenarioRunResult, ItemRunResult
from ._http_client import VevalHttpClient

T = TypeVar("T")


class VevalSdk:
    def __init__(self, options: VevalOptions):
        self._options = options
        self._client = VevalHttpClient(options.api_key, options.endpoint)

    async def run_async(
        self,
        agent_name: str,
        callback: Callable[[VevalExecutionContext], Awaitable[T]],
        input: Any = None,
    ) -> T:
        trace_id = f"tr_{uuid.uuid4().hex}"
        ctx = VevalExecutionContext(trace_id, input)
        started_at = datetime.now(timezone.utc)

        try:
            result = await callback(ctx)
            completed_at = datetime.now(timezone.utc)
            payload = self._build_payload(
                trace_id, agent_name, ctx, input, result,
                "success", None, started_at, completed_at,
            )
            await self._client.send_trace_async(payload)
            return result
        except Exception as ex:
            completed_at = datetime.now(timezone.utc)
            payload = self._build_payload(
                trace_id, agent_name, ctx, input, None,
                "error", str(ex), started_at, completed_at,
            )
            await self._client.send_trace_async(payload)
            raise

    async def get_trace_async(self, trace_id: str) -> Optional[TraceData]:
        return await self._client.get_trace_async(trace_id)

    async def judge_async(
        self,
        criteria: str,
        ctx: VevalExecutionContext,
        model: Optional[str] = None,
    ) -> dict:
        last_step = ctx.steps[-1] if ctx.steps else None
        payload = {
            "criteria": criteria,
            "input": last_step.input if last_step else ctx.input,
            "output": last_step.output if last_step else None,
            "model": model,
        }
        return await self._client.judge_async(payload)

    async def load_snapshot_async(self, trace_id: str) -> Optional[SnapshotData]:
        trace = await self._client.get_trace_async(trace_id)
        return SnapshotData.from_trace(trace) if trace else None

    async def compare_snapshot_async(
        self,
        snapshot_name: str,
        snapshot: SnapshotData,
        ctx: VevalExecutionContext,
    ) -> SnapshotDiff:
        diff = compare_snapshots(snapshot, ctx)
        actual = SnapshotData.from_context(ctx)
        await self._client.post_scenario_run_async(snapshot_name, {
            "passed": not diff.has_changes,
            "pass_count": 0 if diff.has_changes else 1,
            "fail_count": 1 if diff.has_changes else 0,
            "results": [{
                "name": snapshot_name,
                "passed": not diff.has_changes,
                "type": "snapshot",
                "expected": [{"name": s.name, "output": s.output} for s in snapshot.steps],
                "actual": [{"name": s.name, "output": s.output} for s in actual.steps],
                "failures": (
                    [f"added: {s}" for s in diff.added_steps]
                    + [f"removed: {s}" for s in diff.removed_steps]
                    + diff.order_changes
                ),
            }],
        })
        return diff

    async def replay_async(
        self,
        trace: TraceData,
        callback: Callable[[VevalExecutionContext], Awaitable[T]],
        options: Optional[ReplayOptions] = None,
    ) -> ReplayResult:
        options = options or ReplayOptions()
        ctx = VevalExecutionContext(f"tr_{uuid.uuid4().hex}", trace.input)
        if options.mock_llm_responses:
            ctx.load_mock_outputs(trace)

        started_at = datetime.now(timezone.utc)
        output: Any = None
        status = "success"
        error: Optional[str] = None

        try:
            output = await callback(ctx)
        except Exception as ex:
            status = "error"
            error = str(ex)

        completed_at = datetime.now(timezone.utc)
        failures: list[str] = []
        if error:
            failures.append(f"Replay threw exception: {error}")
        for assertion in options.assertions:
            failure = await assertion.evaluate_async(ctx)
            if failure:
                failures.append(failure)

        return ReplayResult(
            failures=failures,
            replayed_context=ctx,
            output=output,
            started_at=started_at,
            completed_at=completed_at,
            status=status,
            error=error,
        )

    async def run_scenario_async(
        self,
        scenario_name: str,
        agent: Callable[[VevalExecutionContext], Awaitable[T]],
        scenario_assertions: list[ITraceAssertion],
        items: Optional[list[ScenarioItem]] = None,
    ) -> ScenarioRunResult:
        if items is None:
            raw_items = await self._client.get_scenario_items_async(scenario_name)
            resolved_items = [
                ScenarioItem(
                    name=i.get("name"),
                    trace_id=i.get("trace_id"),
                    input=i.get("input"),
                )
                for i in raw_items
            ]
        else:
            resolved_items = items

        scenario_result = ScenarioRunResult()

        for item in resolved_items:
            effective_assertions = scenario_assertions + list(item.assertions)
            item_result = ItemRunResult(item=item)

            if item.trace_id:
                trace = await self.get_trace_async(item.trace_id)
                if trace is None:
                    item_result.failures.append(f"trace '{item.trace_id}' not found")
                else:
                    replay_result = await self.replay_async(
                        trace, agent,
                        ReplayOptions(mock_llm_responses=True, assertions=effective_assertions),
                    )
                    item_result.failures.extend(replay_result.failures)
                    item_result.context = replay_result.replayed_context

                    replay_ctx = replay_result.replayed_context
                    if replay_ctx:
                        payload = self._build_payload(
                            replay_ctx.trace_id, scenario_name, replay_ctx,
                            trace.input, replay_result.output,
                            replay_result.status, replay_result.error,
                            replay_result.started_at, replay_result.completed_at,
                            extra_meta={"replay": True, "source_trace_id": item.trace_id},
                        )
                        await self._client.send_trace_async(payload)

            elif item.input is not None:
                ctx = await self._run_and_capture_context(scenario_name, agent, item.input)
                for assertion in effective_assertions:
                    failure = await assertion.evaluate_async(ctx)
                    if failure:
                        item_result.failures.append(failure)
                item_result.context = ctx
            else:
                item_result.failures.append("ScenarioItem must have either trace_id or input")

            scenario_result.results.append(item_result)

        await self._client.post_scenario_run_async(scenario_name, {
            "passed": scenario_result.passed,
            "pass_count": scenario_result.pass_count,
            "fail_count": scenario_result.fail_count,
            "results": [
                {
                    "name": r.item.name or r.item.trace_id or "synthetic",
                    "passed": r.passed,
                    "failures": r.failures,
                }
                for r in scenario_result.results
            ],
        })

        return scenario_result

    async def _run_and_capture_context(
        self,
        agent_name: str,
        agent: Callable[[VevalExecutionContext], Awaitable[T]],
        input: Any,
    ) -> VevalExecutionContext:
        trace_id = f"tr_{uuid.uuid4().hex}"
        ctx = VevalExecutionContext(trace_id, input)
        started_at = datetime.now(timezone.utc)

        try:
            result = await agent(ctx)
            completed_at = datetime.now(timezone.utc)
            payload = self._build_payload(
                trace_id, agent_name, ctx, input, result,
                "success", None, started_at, completed_at,
            )
            await self._client.send_trace_async(payload)
        except Exception as ex:
            completed_at = datetime.now(timezone.utc)
            payload = self._build_payload(
                trace_id, agent_name, ctx, input, None,
                "error", str(ex), started_at, completed_at,
            )
            await self._client.send_trace_async(payload)

        return ctx

    def _build_payload(
        self,
        trace_id: str,
        agent_name: str,
        ctx: VevalExecutionContext,
        input: Any,
        output: Any,
        status: str,
        error: Optional[str],
        started_at: datetime,
        completed_at: datetime,
        extra_meta: Optional[dict] = None,
    ) -> dict:
        steps = _flatten_steps(ctx.steps)
        duration_ms = int((completed_at - started_at).total_seconds() * 1000)
        total_cost = sum(s.cost_usd or 0 for s in steps)

        metadata = dict(ctx.trace_meta)
        if extra_meta:
            metadata.update(extra_meta)

        return {
            "trace_id": trace_id,
            "agent_name": agent_name,
            "input": input,
            "output": output,
            "status": status,
            "error": error,
            "started_at": started_at.isoformat(),
            "completed_at": completed_at.isoformat(),
            "duration_ms": duration_ms,
            "total_cost_usd": total_cost or None,
            "metadata": metadata,
            "steps": [
                {
                    "step_id": s.step_id,
                    "parent_step_id": s.parent_step_id,
                    "name": s.name,
                    "type": s.type,
                    "input": s.input,
                    "output": s.output,
                    "status": s.status,
                    "error": s.error,
                    "started_at": s.started_at.isoformat(),
                    "completed_at": s.completed_at.isoformat() if s.completed_at else None,
                    "duration_ms": s.duration_ms,
                    "cost_usd": s.cost_usd,
                    "tokens_in": s.tokens_in,
                    "tokens_out": s.tokens_out,
                    "model": s.model,
                    "metadata": s.metadata,
                }
                for s in steps
            ],
        }


def _flatten_steps(steps: list[Step]) -> list[Step]:
    result: list[Step] = []
    for s in steps:
        result.append(s)
        result.extend(_flatten_steps(s.children))
    return result

File: node/src/sdk.ts
import { VevalOptions, ResolvedVevalOptions, resolveOptions } from "./options";
import { VevalExecutionContext } from "./context";
import { Step } from "./step";
import { TraceData, StepData, ReplayOptions, ReplayResult } from "./tracing";
import {
  SnapshotData,
  SnapshotDataHelper,
  SnapshotDiff,
  SnapshotComparer,
  SnapshotOptions,
  SnapshotPayloads,
} from "./snapshots";
import { ITraceAssertion, JudgeResult, JudgeOptions } from "./assertions";
import { ScenarioItem, ScenarioRunResult, ItemRunResult } from "./scenarios";
import { VevalHttpClient } from "./http-client";

export class VevalSdk {
  protected readonly opts: ResolvedVevalOptions;
  protected readonly http: VevalHttpClient;

  constructor(options: VevalOptions) {
    this.opts = resolveOptions(options);
    this.http = new VevalHttpClient(this.opts.apiKey, this.opts.endpoint);
  }

  async runAsync<T>(
    agentName: string,
    callback: (ctx: VevalExecutionContext) => Promise<T>,
    input?: unknown
  ): Promise<T> {
    const traceId = `tr_${crypto.randomUUID().replace(/-/g, "")}`;
    const ctx = new VevalExecutionContext(traceId, input ?? null);
    const startedAt = new Date();

    try {
      const result = await callback(ctx);
      const completedAt = new Date();
      const payload = VevalSdk.buildPayload(
        traceId, agentName, ctx,
        input ?? null, result, "success", null, startedAt, completedAt
      );
      await this.http.sendTraceAsync(payload);
      return result;
    } catch (err) {
      const completedAt = new Date();
      const error = err instanceof Error ? err.message : String(err);
      const payload = VevalSdk.buildPayload(
        traceId, agentName, ctx,
        input ?? null, null, "error", error, startedAt, completedAt
      );
      await this.http.sendTraceAsync(payload);
      throw err;
    }
  }

  async getTraceAsync(traceId: string): Promise<TraceData | null> {
    return this.http.getTraceAsync(traceId);
  }

  async judgeAsync(criteria: string, ctx: VevalExecutionContext, options?: JudgeOptions): Promise<JudgeResult> {
    const lastStep = ctx.steps[ctx.steps.length - 1];
    const payload = {
      criteria,
      input: lastStep ? lastStep.input : ctx.input,
      output: lastStep ? lastStep.output : null,
      model: options?.model ?? null,
      threshold: options?.threshold ?? null,
      reference_output: options?.referenceOutput ?? null,
      samples: options?.samples ?? null,
    };
    return this.http.judgeAsync(payload);
  }

  async loadSnapshotAsync(traceId: string): Promise<SnapshotData | null> {
    const trace = await this.http.getTraceAsync(traceId);
    return trace ? SnapshotDataHelper.fromTrace(trace) : null;
  }

  /**
   * Stores a named baseline — from a recorded trace ID (the trace is pinned so retention never deletes it),
   * or from a run you just executed. Saving again under the same name replaces the baseline.
   */
  async saveSnapshotAsync(snapshotName: string, source: string | VevalExecutionContext): Promise<SnapshotData> {
    const payload = typeof source === "string"
      ? SnapshotPayloads.saveFromTrace(snapshotName, source)
      : SnapshotPayloads.saveFromContext(snapshotName, source);
    return this.http.createSnapshotAsync(payload);
  }

  /** The latest stored baseline with this name, or null if none exists. */
  async getSnapshotAsync(snapshotName: string): Promise<SnapshotData | null> {
    return this.http.getSnapshotAsync(snapshotName);
  }

  /** Compares a run against a baseline and records the result in the dashboard. */
  async compareSnapshotAsync(
    snapshotName: string,
    snapshot: SnapshotData,
    ctx: VevalExecutionContext,
    options?: SnapshotOptions
  ): Promise<SnapshotDiff> {
    const diff = SnapshotComparer.compare(snapshot, ctx, options);
    await this.http.postScenarioRunAsync(snapshotName, SnapshotPayloads.run(snapshotName, diff));
    return diff;
  }

  async replayAsync<T>(
    trace: TraceData,
    callback: (ctx: VevalExecutionContext) => Promise<T>,
    options?: Partial<ReplayOptions>
  ): Promise<ReplayResult> {
    const ctx = new VevalExecutionContext(`tr_${crypto.randomUUID().replace(/-/g, "")}`, trace.input);
    if (options?.mock_llm_responses) {
      ctx.loadMockOutputs(trace);
    }

    const startedAt = new Date();
    let output: unknown = null;
    let status = "success";
    let error: string | null = null;

    try {
      output = await callback(ctx);
    } catch (err) {
      status = "error";
      error = err instanceof Error ? err.message : String(err);
    }

    const completedAt = new Date();
    const failures: string[] = [];
    if (error) failures.push(`Replay threw exception: ${error}`);
    for (const assertion of options?.assertions ?? []) {
      const failure = await assertion.evaluate(ctx);
      if (failure) failures.push(failure);
    }

    let recordingDiff: SnapshotDiff | null = null;
    if (options?.compare_with_recording) {
      recordingDiff = SnapshotComparer.compare(SnapshotDataHelper.fromTrace(trace), ctx, options.compare_with_recording);
      if (recordingDiff.has_changes)
        failures.push("Replay drifted from its recording — " + recordingDiff.summary(`recording ${trace.trace_id}`));
    }

    return {
      passed: failures.length === 0,
      failures,
      replayed_context: ctx,
      recording_diff: recordingDiff,
      output,
      started_at: startedAt.toISOString(),
      completed_at: completedAt.toISOString(),
      status,
      error,
    };
  }

  async runScenarioAsync<T>(
    scenarioName: string,
    agent: (ctx: VevalExecutionContext) => Promise<T>,
    scenarioAssertions: ITraceAssertion[],
    items?: ScenarioItem[]
  ): Promise<ScenarioRunResult> {
    const rawItems = items ?? (await this.http.getScenarioItemsAsync(scenarioName)) as ScenarioItem[];
    const results: ItemRunResult[] = [];

    for (const item of rawItems) {
      const effectiveAssertions = [...scenarioAssertions, ...(item.assertions ?? [])];
      const itemResult: ItemRunResult = { item, passed: true, failures: [] };

      if (item.trace_id) {
        const trace = await this.getTraceAsync(item.trace_id);
        if (!trace) {
          itemResult.failures.push(`trace '${item.trace_id}' not found`);
          itemResult.passed = false;
        } else {
          const replayResult = await this.replayAsync(trace, agent, {
            mock_llm_responses: true,
            assertions: effectiveAssertions,
          });
          itemResult.failures.push(...replayResult.failures);
          itemResult.passed = replayResult.passed;
          itemResult.context = replayResult.replayed_context ?? undefined;

          if (replayResult.replayed_context) {
            const replayCtx = replayResult.replayed_context;
            const payload = VevalSdk.buildPayload(
              replayCtx.traceId, scenarioName, replayCtx,
              trace.input, replayResult.output, replayResult.status, replayResult.error,
              new Date(replayResult.started_at), new Date(replayResult.completed_at),
              { replay: true, source_trace_id: item.trace_id }
            );
            await this.http.sendTraceAsync(payload);
          }
        }
      } else if (item.input !== undefined) {
        const ctx = await this._runAndCaptureContext(scenarioName, agent, item.input);
        for (const assertion of effectiveAssertions) {
          const failure = await assertion.evaluate(ctx);
          if (failure) {
            itemResult.failures.push(failure);
            itemResult.passed = false;
          }
        }
        itemResult.context = ctx;
      } else {
        itemResult.failures.push("ScenarioItem must have either trace_id or input");
        itemResult.passed = false;
      }

      results.push(itemResult);
    }

    const passCount = results.filter((r) => r.passed).length;
    const failCount = results.length - passCount;
    const scenarioResult: ScenarioRunResult = {
      passed: failCount === 0,
      pass_count: passCount,
      fail_count: failCount,
      results,
    };

    await this.http.postScenarioRunAsync(scenarioName, {
      passed: scenarioResult.passed,
      pass_count: scenarioResult.pass_count,
      fail_count: scenarioResult.fail_count,
      results: results.map((r) => ({
        name: r.item.name ?? r.item.trace_id ?? "synthetic",
        passed: r.passed,
        failures: r.failures,
        judgments: r.context?.judgments ?? [],
      })),
    });

    return scenarioResult;
  }

  private async _runAndCaptureContext<T>(
    agentName: string,
    agent: (ctx: VevalExecutionContext) => Promise<T>,
    input: unknown
  ): Promise<VevalExecutionContext> {
    const traceId = `tr_${crypto.randomUUID().replace(/-/g, "")}`;
    const ctx = new VevalExecutionContext(traceId, input);
    const startedAt = new Date();

    try {
      const result = await agent(ctx);
      const completedAt = new Date();
      const payload = VevalSdk.buildPayload(
        traceId, agentName, ctx,
        input, result, "success", null, startedAt, completedAt
      );
      await this.http.sendTraceAsync(payload);
    } catch (err) {
      const completedAt = new Date();
      const error = err instanceof Error ? err.message : String(err);
      const payload = VevalSdk.buildPayload(
        traceId, agentName, ctx,
        input, null, "error", error, startedAt, completedAt
      );
      await this.http.sendTraceAsync(payload);
    }

    return ctx;
  }

  static buildPayload(
    traceId: string,
    agentName: string,
    ctx: VevalExecutionContext,
    input: unknown,
    output: unknown,
    status: string,
    error: string | null,
    startedAt: Date,
    completedAt: Date,
    extraMeta?: Record<string, unknown>
  ): TraceData {
    const steps = VevalSdk.flattenSteps(ctx.steps);
    const totalCost = steps.reduce((sum, s) => sum + (s.costUsd ?? 0), 0);

    const metadata: Record<string, unknown> = { ...ctx.traceMeta, ...extraMeta };

    return {
      trace_id: traceId,
      agent_name: agentName,
      input,
      output,
      status,
      error,
      started_at: startedAt.toISOString(),
      completed_at: completedAt.toISOString(),
      duration_ms: completedAt.getTime() - startedAt.getTime(),
      total_cost_usd: totalCost || null,
      metadata,
      steps: VevalSdk.serializeSteps(ctx.steps),
    };
  }

  static flattenSteps(steps: readonly Step[]): Step[] {
    const result: Step[] = [];
    for (const s of steps) {
      result.push(s);
      result.push(...VevalSdk.flattenSteps(s.children));
    }
    return result;
  }

  private static serializeSteps(steps: readonly Step[]): StepData[] {
    const result: StepData[] = [];
    for (const s of steps) {
      result.push({
        step_id: s.stepId,
        parent_step_id: s.parentStepId,
        name: s.name,
        type: s.type,
        input: s.input,
        output: s.output,
        status: s.status,
        error: s.error,
        started_at: s.startedAt.toISOString(),
        completed_at: s.completedAt?.toISOString() ?? null,
        duration_ms: s.durationMs,
        cost_usd: s.costUsd,
        tokens_in: s.tokensIn,
        tokens_out: s.tokensOut,
        model: s.model,
        metadata: { ...s.metadata },
      });
      result.push(...VevalSdk.serializeSteps(s.children));
    }
    return result;
  }
}
